`diff-containers`: Test preconditions, invariants and postconditions for positivity and normality.

jorisdral opened the pull request:
Precursor reference: https://github.com/input-output-hk/anti-diffs/pull/15#discussion_r1098577357

#15 brought back unsafe constructors in the `diff-containers` package. A side-effect of this is that we can not hide the internals (`Data.Map.Diff.Strict.Internal`) behind a truly safe interface (`Data.Map.Diff.Strict`). Instead, downstream code now has the responsibility to use the semi-safe interface faithfully. 

What it means to use diffs in a safe way is now documented in the code, amongst others using PRECONDITION, INVARIANT and POSTCONDITION annotations. This PR also introduces tests for these properties.

The invariant for downstream code to uphold is to only ever apply diffs that are both "normal" and "positive". Faithful use of the API should prevent this from happening.

The README that will be added in resolving #8 will explain both the normality and positivity properties, and it will contain examples of use cases.

dnadales commented on diff-containers/src/Data/Map/Diff/Strict/Internal.hs:
```diff
@@ -199,10 +227,51 @@ null (Diff m) = Map.null m
+
+-- | Check if a diff history is positive, which means it can only contain
+-- @'Insert'@s and @'Delete'@s.
+isPositiveDiffHistory :: DiffHistory v -> Bool
```
Consider adding `isPositiveDiffHistory` and `isNormalDiffHistory` as local definitions if we only use them in one function.

jorisdral replied:
The tests use these definitions, and I think it would be useful to keep exposing it to the user even though it is an Internal module

dnadales commented on diff-containers/src/Data/Map/Diff/Strict/Internal.hs:
```diff
@@ -199,10 +227,51 @@ null (Diff m) = Map.null m
+-- normal form.
+isNormalDiffHistory :: Eq v => DiffHistory v -> Bool
+isNormalDiffHistory (DiffHistory vs) =
+    snd $ foldl' f (Nothing, True) vs
```
Would something like this work? 

```haskell
    all $ fmap (not . uncurry areInverses) (consecutivePairs vs)

consecutivePairs [] = []
consecutivePairs xs = zip xs (tail ys)
```

jorisdral replied:
Zipping would probably be better, yes, so I'll use your suggestion!

dnadales commented on diff-containers/src/Data/Map/Diff/Strict/Internal.hs:
```diff
@@ -309,6 +414,10 @@ applyDiffForKeys m ks (Diff diffs) =
+--
+-- PRECONDITION: The diff that is to be applied is both normal and positive.
+--
+-- POSTCONDITION: No error is thrown.
```
I wonder if `No error is thrown` is a postcondition :thinking:

jorisdral replied:
Do you mean as in: the postcondition will not hold, or its formulation is not really a postcondition? 

For the former, I just realised that the postcondition is not specific enough: we only guarantee not throwing the "applyDiff failed" error, but of course any other error could still be thrown.  

For the latter, I'm not sure. Can postconditions not mention errors? Would you have a suggestion for an alternative to name postcondition?

dnadales replied:
It's just that I'm used to postconditions being predicates on inputs and outputs. Exceptions break this rule, but I don't really have a better alternative (other than mention `No error is thrown` in the end of the comment, not preceded by any section name (like `POSTCONDITION`).

dnadales commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -43,123 +46,219 @@ import           Test.Util
+            prop_diffingIsNormal @(OftenSmall Int) @(OftenSmall Int)
+        ]
+    , testGroup "Preserving positivity and normality" [
+          localOption (QuickCheckTests 1000) $
```
Are we missing a `0` here or is this prop more expensive to test?

jorisdral replied:
It's more expensive. I try to keep the running time of most tests within a few seconds

dnadales commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -43,123 +46,219 @@ import           Test.Util
-prop_applyAllAndApplySum m ds =
-  all isPositive ds ==> foldlM applyDiff m ds === applyDiff m (mconcat ds)
+prop_applyAllAndApplySum m (fmap getNPDiff -> ds) =
+  all isPositive ds && all isNormal ds ==> foldlM applyDiff m ds === applyDiff m (mconcat ds)
```
Maybe it's worth tagging how many cases are we filtering here (due to the `==>` operator).

dnadales replied:
And the same for other uses of `==>` below.

Jasagredo replied:
As these are "guaranteed" by the generator, I would actually either test the generators (the thing to do :tm:) or change this to an `assert` that will crash if this is not true.

jorisdral replied:
> Maybe it's worth tagging how many cases are we filtering here (due to the `==>` operator).

QuickCheck will already report how many tests are discarded, is that 

> As these are "guaranteed" by the generator, I would actually either test the generators (the thing to do :tm:) or change this to an `assert` that will crash if this is not true.

I added the implications to see whether the generated values were actually normal/positive, so that if the tests are discarded, I could see that my generators are wrong. Testing the generators and adding assertions is probably better :P

dnadales commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -174,3 +273,78 @@ instance Arbitrary v => Arbitrary (DiffEntry v) where
+instance (Ord k, Eq v, Arbitrary k, Arbitrary v)
+      => Arbitrary (NormalDiff k v) where
+  arbitrary             = NormalDiff <$> arbitrary `suchThat` isNormal
+  shrink (NormalDiff d) = NormalDiff <$> shrink d
```
Can't `shrink d` result in non-normal diff sequences? It seems so.

jorisdral replied:
You are absolutely right, I will change this

dnadales commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -174,3 +273,78 @@ instance Arbitrary v => Arbitrary (DiffEntry v) where
+  shrink (PositiveDiffHistory h) = PositiveDiffHistory <$> shrink h
+
+{------------------------------------------------------------------------------
+  Modifiers: normality
```
Perhaps we want to test the generators in this section.

jorisdral replied:
Will do

dnadales commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -174,3 +273,78 @@ instance Arbitrary v => Arbitrary (DiffEntry v) where
+instance (Ord k, Eq v, Arbitrary k, Arbitrary v)
+      => Arbitrary (NPDiff k v) where
+  arbitrary             = NPDiff <$> arbitrary `suchThat` (\d -> isNormal d && isPositive d)
+  shrink (NPDiff d) = NPDiff <$> shrink d
```
Usually the use of `suchThat` in the generator requires special attention in the shink function.

dnadales commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -174,3 +273,78 @@ instance Arbitrary v => Arbitrary (DiffEntry v) where
+  deriving newtype (Semigroup, Monoid, Group)
+
+instance (Arbitrary v, Eq v) => Arbitrary (NPDiffHistory v) where
+  arbitrary = NPDiffHistory <$> arbitrary `suchThat` (\d -> isNormalDiffHistory d && isPositiveDiffHistory d)
```
Same here: we should probably check that the shrinking function preserves the invariant.

Jasagredo commented on diff-containers/src/Data/Map/Diff/Strict.hs:
```diff
@@ -1,3 +1,20 @@
+{-# OPTIONS_GHC -Wno-unused-imports #-}
```
Left-over?

jorisdral replied:
The imports of `Data.Map.Strict` and `Data.Group` are unused, but they are useful for hyperlinking in the haddock comments, so this option is still necessary

Jasagredo commented on diff-containers/src/Data/Map/Diff/Strict/Internal.hs:
```diff
@@ -199,10 +227,51 @@ null (Diff m) = Map.null m
+
+-- | A normal diff has resolved all sums and subtractions internally. Applying a
+-- non-normal diff is considered unsafe.
+isNormal :: Eq v => Diff k v -> Bool
```
I guess normal here stands for `normal form`?

jorisdral replied:
Yup, I will be more specific about that

Jasagredo commented on diff-containers/test/Test/Data/Map/Diff/Strict.hs:
```diff
@@ -43,123 +46,219 @@ import           Test.Util
+        , localOption (QuickCheckTests 10000) $
+          testProperty "prop_summingDiffsPreservesNormality" $
+            prop_summingDiffsPreservesNormality @(OftenSmall Int) @(OftenSmall Int)
+        , localOption (QuickCheckTests 100000) $
```
Same question here? Is this running more than the others for some reason?

jorisdral replied:
It's a less expensive test, so I run it more. 

In testing `diff-containers`, I experienced that running too few tests would often not hit interesting test cases that ended in test failure. I've tried to counter this by making the generators generate test cases from a smaller set (see `OftenSmall`), but even then I did not want to make the generators too specific, so I run more tests

jorisdral commented:
@dnadales @Jasagredo I have added tests for the generators/shrinkers. While writing them, I also realised that "positivity implies normality": a positive diff or diff history is by definition also normal. So I've added a test for that as well, and the documentation now also mentions this

dnadales reviewed (approved):
Thank you! I'm really happy with the tests for the generators
